security/src/main/scala/pl/touk/nussknacker/ui/security/api/AuthenticationManager.scala

```diff
@@ -0,0 +1,107 @@
+      authenticateWithImpersonation(impersonatingUserAuthCredentials, impersonatedUserIdentity)
+  }
+
+  def authenticationEndpointInput(): EndpointInput[AuthCredentials] =
+    authenticationResources
+      .authenticationMethod()
+      .and(impersonationHeaderEndpointInput)
+      .map(mappedAuthenticationEndpointInput)
+
+  private def provideOrImpersonateUser(user: AuthenticatedUser): Directive1[AuthenticatedUser] = {
+    optionalHeaderValueByName(impersonateHeaderName).flatMap {
+      case Some(impersonatedUserIdentity) =>
+        authenticationResources.getImpersonatedUserData(impersonatedUserIdentity) match {
+          case Some(impersonatedUserData) =>
+            provide(AuthenticatedUser.createImpersonatedUser(user, impersonatedUserData))
+          case None => reject
+        }
+      case None => provide(user)
+    }
+  }
+
+  private def authenticateWithImpersonation(
+      impersonatingUserAuthCredentials: PassedAuthCredentials,
+      impersonatedUserIdentity: ImpersonatedUserIdentity,
+  ): Future[Option[AuthenticatedUser]] = (for {
+    impersonatingUser <- OptionT(authenticationResources.authenticate(impersonatingUserAuthCredentials))
+    impersonatedUserData <- OptionT(
+      Future { authenticationResources.getImpersonatedUserData(impersonatedUserIdentity.value) }
+    )
+  } yield AuthenticatedUser.createImpersonatedUser(impersonatingUser, impersonatedUserData)).value
+
+  private def handleAuthorizationFailedRejection: Directive0 = handleRejections(
+    RejectionHandler
+      .newBuilder()
+      // If the authorization rejection was caused by anonymous access,
+      // we issue the Unauthorized status code with a challenge instead of the Forbidden
+      .handle { case AuthorizationFailedRejection => authenticationResources.authenticate() { _ => reject } }
+      .result()
+  )
+
+  private def authenticateWithAnonymousAccess(): Future[Option[AuthenticatedUser]] = Future {
+    authenticationResources.getAnonymousRole.map(role => AuthenticatedUser.createAnonymousUser(Set(role)))
+  }
+
+}
+
+object AuthenticationManager {
+  val impersonateHeaderName = "Impersonate-User-Identity"
+  val impersonationHeaderEndpointInput: EndpointIO.Header[Option[String]] =
+    header[Option[String]](impersonateHeaderName)
+
+  val mappedAuthenticationEndpointInput: Mapping[(Option[String], Option[String]), AuthCredentials] =
+    Mapping.fromDecode[(Option[String], Option[String]), AuthCredentials] {
+      case (Some(passedCredentials), None) => DecodeResult.Value(PassedAuthCredentials(passedCredentials))
+      case (Some(passedCredentials), Some(identity)) =>
+        DecodeResult.Value(
+          ImpersonatedAuthCredentials(PassedAuthCredentials(passedCredentials), ImpersonatedUserIdentity(identity))
+        )
+      case (None, None) => DecodeResult.Value(NoCredentialsProvided)
+      case _            => DecodeResult.Missing
```

Comment:
let's do it explicity: `(None, Some(_))`

It means that the creds were not passed, but there was the impersonation header present. We return `Missing`. Maybe we should add a comment about why here we return Missing instead of `NoCredentialsProvided`. It's not obvious at first glance. 

WDYT?

Reply:
I added it explicitly and commented why we return that.